// node-iframe/dist/iframe.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.fetchFrame = exports.configureCacheControl = exports.configureTemplates = exports.configureResourceControl = exports.appSourceConfig = void 0;
var is_url_1 = __importDefault(require("is-url"));
var isomorphic_unfetch_1 = __importDefault(require("isomorphic-unfetch"));
var cheerio_1 = require("cheerio");
var templates_1 = require("./templates");
exports.configureTemplates = templates_1.configureTemplates;
var config_1 = require("./config");
var cache_1 = require("./cache");
exports.configureCacheControl = cache_1.configureCacheControl;
var appSourceConfig = config_1.defaultConfig;
exports.appSourceConfig = appSourceConfig;
var mutateSource = function (_a, url, $html) {
    var _b = _a.src, src = _b === void 0 ? "" : _b, key = _a.key;
    return __awaiter(void 0, void 0, void 0, function () {
        var res, source, e_1;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    if (!(src && src[0] === "/")) return [3, 6];
                    _c.label = 1;
                case 1:
                    _c.trys.push([1, 5, , 6]);
                    return [4, isomorphic_unfetch_1["default"](url + "/" + src, {
                            headers: config_1.headers
                        })];
                case 2:
                    res = _c.sent();
                    return [4, res.text()];
                case 3:
                    source = _c.sent();
                    return [4, $html(key).html(source)];
                case 4:
                    _c.sent();
                    return [3, 6];
                case 5:
                    e_1 = _c.sent();
                    console.error(e_1);
                    return [3, 6];
                case 6: return [2];
            }
        });
    });
};
function renderErrorHtml(_a) {
    var url = _a.url, server = _a.server, _b = _a.noPage, noPage = _b === void 0 ? false : _b;
    return Object.assign(cheerio_1.load(!url
        ? templates_1.templateModel[templates_1.TemplateType.error]
        : templates_1.templateModel[templates_1.TemplateType.notFound]), server ? { status: Number("" + 40 + (!url || noPage ? 4 : 0)) } : {});
}
function renderHtml(_a, server) {
    var e_2, _b, e_3, _c, e_4, _d;
    var url = _a.url, baseHref = _a.baseHref, config = _a.config;
    if (server === void 0) { server = false; }
    return __awaiter(this, void 0, void 0, function () {
        var cachedHtml, e_5, _e, inline, cors, response, html, $html, inlineMutations_2, _f, _g, key, attribute_1, e_2_1, inlineMutations_1, inlineMutations_1_1, com, key, attribute, src, element, e_3_1, _h, _j, key, e_4_1, e_6;
        return __generator(this, function (_k) {
            switch (_k.label) {
                case 0:
                    if (!is_url_1["default"](url)) {
                        return [2, renderErrorHtml({ url: url, server: server })];
                    }
                    _k.label = 1;
                case 1:
                    _k.trys.push([1, 3, , 4]);
                    return [4, cache_1.appCache.get(url)];
                case 2:
                    cachedHtml = _k.sent();
                    if (cachedHtml) {
                        return [2, cheerio_1.load(cachedHtml)];
                    }
                    return [3, 4];
                case 3:
                    e_5 = _k.sent();
                    console.error(e_5);
                    return [3, 4];
                case 4:
                    _e = {
                        inline: __assign(__assign({}, appSourceConfig.inline), config === null || config === void 0 ? void 0 : config.inline),
                        cors: __assign(__assign({}, appSourceConfig.cors), config === null || config === void 0 ? void 0 : config.cors)
                    }, inline = _e.inline, cors = _e.cors;
                    _k.label = 5;
                case 5:
                    _k.trys.push([5, 50, , 51]);
                    return [4, isomorphic_unfetch_1["default"](url, {
                            headers: config_1.headers
                        })];
                case 6:
                    response = _k.sent();
                    return [4, response.text()];
                case 7:
                    html = _k.sent();
                    $html = cheerio_1.load(html);
                    if (!!!baseHref) return [3, 9];
                    return [4, $html("head").prepend("<base target=\"_self\" href=\"" + url + "\">")];
                case 8:
                    _k.sent();
                    _k.label = 9;
                case 9:
                    inlineMutations_2 = [];
                    _k.label = 10;
                case 10:
                    _k.trys.push([10, 16, 17, 22]);
                    _f = __asyncValues(Object.keys(inline));
                    _k.label = 11;
                case 11: return [4, _f.next()];
                case 12:
                    if (!(_g = _k.sent(), !_g.done)) return [3, 15];
                    key = _g.value;
                    if (!inline[key]) return [3, 14];
                    attribute_1 = "src";
                    return [4, $html(key).attr(attribute_1, function (_, src) {
                            if (src) {
                                inlineMutations_2.push({ key: key, attribute: attribute_1, src: src });
                            }
                            return src;
                        })];
                case 13:
                    _k.sent();
                    _k.label = 14;
                case 14: return [3, 11];
                case 15: return [3, 22];
                case 16:
                    e_2_1 = _k.sent();
                    e_2 = { error: e_2_1 };
                    return [3, 22];
                case 17:
                    _k.trys.push([17, , 20, 21]);
                    if (!(_g && !_g.done && (_b = _f["return"]))) return [3, 19];
                    return [4, _b.call(_f)];
                case 18:
                    _k.sent();
                    _k.label = 19;
                case 19: return [3, 21];
                case 20:
                    if (e_2) throw e_2.error;
                    return [7];
                case 21: return [7];
                case 22:
                    _k.trys.push([22, 29, 30, 35]);
                    inlineMutations_1 = __asyncValues(inlineMutations_2);
                    _k.label = 23;
                case 23: return [4, inlineMutations_1.next()];
                case 24:
                    if (!(inlineMutations_1_1 = _k.sent(), !inlineMutations_1_1.done)) return [3, 28];
                    com = inlineMutations_1_1.value;
                    key = com.key, attribute = com.attribute, src = com.src;
                    element = key + "[" + attribute + "=\"" + src + "\"]";
                    return [4, mutateSource({ key: element, src: src }, url, $html)];
                case 25:
                    _k.sent();
                    return [4, $html(element).removeAttr(attribute)];
                case 26:
                    _k.sent();
                    _k.label = 27;
                case 27: return [3, 23];
                case 28: return [3, 35];
                case 29:
                    e_3_1 = _k.sent();
                    e_3 = { error: e_3_1 };
                    return [3, 35];
                case 30:
                    _k.trys.push([30, , 33, 34]);
                    if (!(inlineMutations_1_1 && !inlineMutations_1_1.done && (_c = inlineMutations_1["return"]))) return [3, 32];
                    return [4, _c.call(inlineMutations_1)];
                case 31:
                    _k.sent();
                    _k.label = 32;
                case 32: return [3, 34];
                case 33:
                    if (e_3) throw e_3.error;
                    return [7];
                case 34: return [7];
                case 35: return [4, $html("[src=\"undefined\"]").removeAttr("src")];
                case 36:
                    _k.sent();
                    _k.label = 37;
                case 37:
                    _k.trys.push([37, 43, 44, 49]);
                    _h = __asyncValues(Object.keys(cors));
                    _k.label = 38;
                case 38: return [4, _h.next()];
                case 39:
                    if (!(_j = _k.sent(), !_j.done)) return [3, 42];
                    key = _j.value;
                    if (!cors[key]) return [3, 41];
                    return [4, $html(key).attr("crossorigin", cors[key])];
                case 40:
                    _k.sent();
                    _k.label = 41;
                case 41: return [3, 38];
                case 42: return [3, 49];
                case 43:
                    e_4_1 = _k.sent();
                    e_4 = { error: e_4_1 };
                    return [3, 49];
                case 44:
                    _k.trys.push([44, , 47, 48]);
                    if (!(_j && !_j.done && (_d = _h["return"]))) return [3, 46];
                    return [4, _d.call(_h)];
                case 45:
                    _k.sent();
                    _k.label = 46;
                case 46: return [3, 48];
                case 47:
                    if (e_4) throw e_4.error;
                    return [7];
                case 48: return [7];
                case 49:
                    cache_1.appCache.set(url, $html.html());
                    if (server) {
                        $html.status = 200;
                    }
                    return [2, $html];
                case 50:
                    e_6 = _k.sent();
                    console.error(e_6);
                    return [3, 51];
                case 51: return [2, renderErrorHtml({ url: url, server: server, noPage: true })];
            }
        });
    });
}
function fetchFrame(model) {
    return __awaiter(this, void 0, void 0, function () {
        var $html;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4, renderHtml(model, typeof process !== "undefined")];
                case 1:
                    $html = _a.sent();
                    return [2, $html === null || $html === void 0 ? void 0 : $html.html()];
            }
        });
    });
}
exports.fetchFrame = fetchFrame;
function configureResourceControl(appConfig) {
    exports.appSourceConfig = appSourceConfig = Object.assign({}, config_1.defaultConfig, {
        cors: __assign(__assign({}, config_1.defaultCorsConfig), appConfig === null || appConfig === void 0 ? void 0 : appConfig.cors),
        inline: __assign(__assign({}, config_1.defaultInlineConfig), appConfig === null || appConfig === void 0 ? void 0 : appConfig.inline)
    });
}
exports.configureResourceControl = configureResourceControl;
function createIframe(req, res, next) {
    var _this = this;
    res.createIframe = function (model) { return __awaiter(_this, void 0, void 0, function () {
        var $html;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4, renderHtml(model, true)];
                case 1:
                    $html = _a.sent();
                    res.status(($html === null || $html === void 0 ? void 0 : $html.status) || 200).send($html.html());
                    return [2];
            }
        });
    }); };
    next();
}
exports["default"] = createIframe;
